### DISCUSSION TITLE
Compiling plugins

### DISCUSSION THREAD

**Discussion by richardmonette** [libvips/libvips#3331] [General]
I'm working through trying to compile a vips plugin and getting a bit mixed up between the examples I've seen online. I'm using vips-8.10.5 and Ubuntu 21.10.

I've read through http://libvips.blogspot.com/2014/01/how-to-write-operation-in-vips8.html and then taken a look at the accompanying code here https://gist.github.com/jcupitt/8452027#file-try208-c-L183. I was able to compile and run this example, however it seems like what's going on here is that the compiled output is an executable which uses the negative plugin, rather than a dynamic library / .so file etc. which could be dropped into the vips plugin directory.

With the goal of making a plugin I can use with an existing install of vips I have been reading through https://www.libvips.org/API/current/extending.html which suggests "use GModule to make a binary plugin that will be loaded by libvips at startup"

I took a look at https://github.com/jcupitt/vips-gmic/blob/master/vips_gmic.cpp#L365 which seems to show exporting the module as well as the https://github.com/jcupitt/vips-gmic/blob/master/vips_gmic.cpp#L393 where gmodule seems too handle the `vips_gmic_get_type()` call.

I'm looking now at modifying the https://github.com/jcupitt/vips-gmic/blob/master/Makefile to work with a minimal version of the negative example but without a main to generate a plugin.

The negative example uses .c, but the gmic examples uses .cpp, is cpp necessary to make a plugin?

**Comment by kleisauke** [comment]
C++ shouldn't be necessary to make a dynamic loadable module, it should work for plain old C too. Did you see?:
https://github.com/kleisauke/vips-gmic

That simplifies the vips-gmic example and updates it for the latest libvips and G'MIC. Though, I'm not sure if it still buildable.

FWIW, I once had a idea to create another repository to demonstrate libvips' module system. Liquid rescale / seam carving seemed fun, but it looks like it's currently patent-encumbered, see e.g.: https://patents.google.com/patent/US8213745.

**Comment by richardmonette** [comment]
Hrm, I made a bit more progress, but seem to be getting an error now about the "package_table"

```
% make debug                                                                                
cc  `pkg-config vips-cpp --cflags`  -c -o vips_negative.o vips_negative.c
g++ -o vips-negative.plg -shared vips_negative.o
% vips -p ./vips-negative.plg 
vips_negative: module init
plugin: unable to find symbol "package_table" in plugin "./vips-negative.plg"
plugin: 'package_table': ./vips-negative.plg: undefined symbol: package_table
```

**Reply by richardmonette** [reply]
🤔 so I can see where vips is trying to find the "package_table" https://github.com/libvips/libvips/blob/f11b58ff6c602549d553c57e506142de5b6d28e9/libvips/deprecated/package.c#L698

https://docs.gtk.org/gmodule/method.Module.symbol.html it appears to be trying to "Gets a symbol pointer from a module, such as one exported by G_MODULE_EXPORT"

Bit stumped as to where that should have been exported from though, not seeing it obviously by name in any of the examples

**Reply by kleisauke** [reply]
The `--plugin` / `-p` option only loads vips8 plugins when building with `-Ddeprecated=false`, see:
https://github.com/libvips/libvips/blob/aa439094903861bed6fe387b331761ce03fc5855/tools/vips.c#L742-L753

I think the recommended way is to just install the dynamically loadable module in `$VIPSHOME/lib/vips-plugins-x.y` or `$VIPSHOME/lib/vips-modules-x.y`, for example:
https://gist.github.com/kleisauke/0b73722fba49f491436a8364993db9f3

**Reply by richardmonette** [reply]
@kleisauke Aha! That was it! I've got it working now. Thank you very much for your help.

**Comment by richardmonette** [comment]
@jcupitt I haven't been able to find any other mention of "package_table" other than in this older post https://github.com/libvips/libvips/issues/96 which links to some code https://pastebin.com/UZ0vWJLb which does have something like 

```
im_package package_table = {
  "layer",
  IM_NUMBER(function_list),
  function_list
};
```

defined. Is this still necessary? I don't see that in any other examples online.
